### ISSUE TITLE
Quantized model - how to properly save and use QuantizationSimModel

### ISSUE THREAD

**Issue by mstaczek** [quic/aimet#2199]
Hi,

In short: I would like to quantize a torch ResNet18 with AdaRound, save it and later run some benchmarks/experiments with it. How to save/export a model after quantization for later usage? 

I found:
1. Documentation about [aimet_torch.quantsim](https://quic.github.io/aimet-pages/releases/1.25.0/api_docs/torch_quantsim.html#aimet_torch.quantsim.save_checkpoint):
- `quantsim.save_checkpoint` and `quantsim.load_checkpoint` are said to 'save and restore the quantized model',
- it works and lets me save and load a `QuantizationSimModel` object,
2. Documentation about [aimet_torch.quantsim.QuantizationSimModel](https://quic.github.io/aimet-pages/releases/1.25.0/api_docs/torch_quantsim.html#aimet_torch.quantsim.QuantizationSimModel):
- it does export a PyTorch model but also JSON/YAML files with some calibration parameters (_encodings_),
3. Answer for previous _aimet_ version: [here](https://github.com/quic/aimet/issues/88#issuecomment-714226118):
> (...) The exported model from AIMET, which you got from sim.export() call, is a FP32 model that has been optimized to run on quantized runtimes. (...)

My doubts are:
1. Is the `pth` model from `sim.export()` quantized despite being in FP32? What does '_has been optimized to run on quantized runtimes_' mean?
2. Is it somehow 'visible' by looking at model's weights that they were quantized? Their type remains the same but values have slightly changed which slighly reduced the accuracy.
3. The `pth` model from `sim.export()` has the same size as the original ResNet18 saved with `torch.save()`. I assume it is caused by weigths still having the same type (number of bits) as the original model. I expected the quantized model (at least theoretically) to have a smaller size but I see it is not the case in reality.

Thank you for all explanations!

**Comment by quic-hitameht** [comment]
@mstaczek here are answers to questions

1. If you have only created `QuantizationSimModel` simulation model from original FP32 model and not used any post training quantization technique (like Adaround), the pth from `sim.export()` will have same weights as FP32 model. But, when you apply Adaround first to obtain updated model(optimized to run on quantized runtime) to then pass into `QuantizationSimModel` to find quantized accuracy.
2. `sim.export()` saves copy of the model with quantization nodes removed along with encodings JSON file which contains quantization scale/offset parameters for each activations and weights in the model. This JSON file can be ingested by the hardware and will match with the exported model. Here is an example on how to apply Adaround.  https://github.com/quic/aimet/blob/develop/Examples/torch/quantization/adaround.ipynb
3. As I said, .pth from `sim.export()` will have the same size as original FP32 model but will also save encodings JSON file to be ingested by the hardware.

Let me know if these answered your questions.

**Comment by mstaczek** [comment]
Hi @quic-hitameht, 

I did try and managed to run the Example notebook that you have mentioned: 
[https://github.com/quic/aimet/blob/develop/Examples/torch/quantization/adaround.ipynb]( https://github.com/quic/aimet/blob/develop/Examples/torch/quantization/adaround.ipynb)

My question concerns the last cell of this tutorial, that runs:
> dummy_input = dummy_input.cpu()
sim.export(path='./output/', filename_prefix='resnet18_after_adaround', dummy_input=dummy_input)

Based on your answers I understand that a file `resnet18_after_adaround.pth` is the quantized version of the model but I am not sure about this JSON. In short I would like to use the quantized model for inference on new data so:
a) should I save the whole `sim` object and use `sim.model(data)` to make new predictions, or
b) should I use `model = torch.load('resnet18_after_adaround.pth')` and then `model(data)`.

Lastly, thank you for the answers. I know it may sound silly but I expected the change in weigths to be visible like `0.123456` -> `0.123000` despite the change happening in binary and not decimal ;)

**Comment by quic-hitameht** [comment]
You should be able to find `resnet18_after_adaround.encodings` JSON file under `./output/` directory which has quantization scale/offsets parameters needed along with `resnet18_after_adaround.pth` to run it on the hardware.

Now, if you want to just evaluate model's performance off-target (simulate quantized accuracy), you should always use `sim.model(data)` for given data because your `sim.model` has AIMET added quantization simulation ops in-place. You can also do `print(sim)` to find that information.

**Comment by mstaczek** [comment]
Thank you for your patience and explanations!

So I will keep using `sim.model(data)` for making new predictions and evaluating quantized model performance. Thanks and good job!
